rust/tests/test_write_resiliency.rs

```diff
@@ -68,78 +72,156 @@ mod test {
+        replace_dn_on_failure_conf.insert(
+            "dfs.client.block.write.replace-datanode-on-failure.enable".to_string(),
+            "true".to_string(),
+        );
+        replace_dn_on_failure_conf.insert(
+            "dfs.client.block.write.replace-datanode-on-failure.policy".to_string(),
+            "ALWAYS".to_string(),
+        );
+
+        let _dfs = MiniDfs::with_features(&HashSet::from([DfsFeatures::HA]));
+        let client = Client::default_with_config(replace_dn_on_failure_conf).unwrap();
+
+        let file = "/testfile_replace_failed_datanode";
         let bytes_to_write = 2usize * 1024 * 1024;
 
         let mut data = BytesMut::with_capacity(bytes_to_write);
         for i in 0..(bytes_to_write / 4) {
             data.put_i32(i as i32);
         }
-
-        // Test connection failure before writing data
-        let mut writer = client
-            .create(file, WriteOptions::default().replication(3))
-            .await?;
-
-        WRITE_CONNECTION_FAULT_INJECTOR.store(true, Ordering::SeqCst);
-
         let data = data.freeze();
-        writer.write(data.clone()).await?;
-        writer.close().await?;
-
-        let reader = client.read("/testfile").await?;
-        check_file_content(&reader, data.clone()).await?;
 
-        // Test connection failure after data has been written
         let mut writer = client
-            .create(file, WriteOptions::default().replication(3).overwrite(true))
+            .create(file, WriteOptions::default().replication(3))
             .await?;
 
-        writer.write(data.slice(..bytes_to_write / 2)).await?;
-
-        // Give a little time for the packets to send
-        tokio::time::sleep(Duration::from_millis(100)).await;
+        writer.write(data.slice(..bytes_to_write / 3)).await?;
 
         WRITE_CONNECTION_FAULT_INJECTOR.store(true, Ordering::SeqCst);
-
-        writer.write(data.slice(bytes_to_write / 2..)).await?;
-        writer.close().await?;
-
-        let reader = client.read("/testfile").await?;
-        check_file_content(&reader, data.clone()).await?;
-
-        // Test failure in from ack status before any data is written
-        let mut writer = client
-            .create(file, WriteOptions::default().replication(3).overwrite(true))
-            .await?;
-
-        *WRITE_REPLY_FAULT_INJECTOR.lock().unwrap() = Some(2);
-
-        writer.write(data.clone()).await?;
-        writer.close().await?;
-
-        let reader = client.read("/testfile").await?;
-        check_file_content(&reader, data.clone()).await?;
-
-        // Test failure in from ack status after some data has been written
-        let mut writer = client
-            .create(file, WriteOptions::default().replication(3).overwrite(true))
+        writer
+            .write(data.slice(bytes_to_write / 3..2 * bytes_to_write / 3))
             .await?;
+        WRITE_CONNECTION_FAULT_INJECTOR.store(false, Ordering::SeqCst);
```

Comment:
If you replace this (not needed because the value is reset to false when grabbed in the block writer) with
```rust
tokio::time::sleep(Duration::from_millis(100)).await;
```

It can make this test fail. The sleep is needed since the actual packet writing is in a different task, it might not actually trigger a failure right away before the next write below.

Still trying to figure out why, looks like a mismatch between the number of bytes we thought we have successfully written and what the datanode thinks

Reply:
Think I figured it out https://github.com/Kimahriman/hdfs-native/pull/214

Reply:
Sorry for the late reply, thanks for explaining and fixing this issue